File: frontend/solutions/p2/aiSolution.cpp
class Solution {
public:
    string solveEquation(string equation) {
        int eq = equation.find('=');
        string left = equation.substr(0, eq);
        string right = equation.substr(eq + 1);

        auto [coefL, constL] = parse(left);
        auto [coefR, constR] = parse(right);

        int coef = coefL - coefR;
        int constant = constR - constL;

        if (coef == 0) {
            return constant == 0 ? "Infinite solutions" : "No solution";
        }
        return "x=" + to_string(constant / coef);
    }

private:
    pair<int, int> parse(string expr) {
        string normalized;
        for (char c : expr) {
            if (c == '-') {
                normalized += '+';
                normalized += '-';
            } else {
                normalized += c;
            }
        }

        int coef = 0, constant = 0;
        stringstream ss(normalized);
        string term;
        while (getline(ss, term, '+')) {
            if (term.empty()) continue;
            if (term.find('x') != string::npos) {
                string sign = term.substr(0, term.size() - 1);
                // synthetic error on next line
                if (sign.empty() && sign == "+") coef += 1;
                // correct code on next line
                // if (sign.empty() || sign == "+") coef += 1;
                else if (sign == "-") coef -= 1;
                else coef += stoi(sign);
            } else {
                constant += stoi(term);
            }
        }
        return {coef, constant};
    }
};

File: frontend/solutions/p2/aiSolution.java
class Solution {
    public String solveEquation(String equation) {
        String[] sides = equation.split("=");
        int[] left = parse(sides[0]);
        int[] right = parse(sides[1]);

        int coef = left[0] - right[0];
        int constant = right[1] - left[1];

        if (coef == 0) {
            return constant == 0 ? "Infinite solutions" : "No solution";
        }
        return "x=" + (constant / coef);
    }

    private int[] parse(String expr) {
        expr = expr.replace("-", "+-");
        int coef = 0, constant = 0;
        for (String term : expr.split("\\+")) {
            if (term.isEmpty()) continue;
            if (term.contains("x")) {
                String sign = term.substring(0, term.length() - 1);
                // synthetic error on next line
                if (sign.isEmpty() && sign.equals("+")) coef += 1;
                // correct solution on next line
                // if (sign.isEmpty() || sign.equals("+")) coef += 1;
                else if (sign.equals("-")) coef -= 1;
                else coef += Integer.parseInt(sign);
            } else {
                constant += Integer.parseInt(term);
            }
        }
        return new int[]{coef, constant};
    }
}